File: 2020/03_TobogganTrajectory/trees.lua
-- ======================================================================
-- Toboggan Trajectory
--   Advent of Code 2020 Day 03 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         t r e e s . l u a
-- ======================================================================
-- "A solver for the Advent of Code 2020 Day 03 puzzle"

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Trees = { part2 = false, text = {}, rows = 0, cols = 0 }
-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local TREE = '#'

-- ======================================================================
--                                                                  Trees
-- ======================================================================

-- "Object for Toboggan Trajectory"

function Trees:Trees (o)

  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.rows = o.rows or 0
  o.cols = o.cols or 0

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Trees:_process_text(text)
  -- "Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)

  -- 1. Set the number of rows and columns
  self.rows = #text
  self.cols = #text[1]

end

function Trees:is_tree(col, row)
  -- Returns true if there is a tree at (col, row)
  
  -- 1. There are no trees below the forest
  if self:is_below(row) then
    return false
  end
  
  -- 2. Check the forest square
  mcol = 1 + (col - 1) % self.cols
  if TREE == self.text[row]:sub(mcol, mcol) then
    return true
  end
  return false
end

function Trees:is_below(row)
  -- Returns true if beyond forest
  return row > self.rows
end

function Trees:count_trees(delta_col, delta_row)
  -- Returns number of trees hit in toboggen ride

  -- 1. Start at the very beginning
  col, row, count = 1, 1, 0
  
  -- 2. Loop until we are out of the forest
  repeat
    
    -- 3. If this is a tree, count it
    if self:is_tree(col, row) then
      count = count + 1
    end
    
    -- 4. Advance to next location
    col = col + delta_col
    row = row + delta_row
    
  until self:is_below(row)
  
  -- 5. Return the number of tree hit
  return count
end

  
function Trees:part_one(args)
  -- "Returns the solution for part one"

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  return self:count_trees(3, 1)
end


function Trees:part_two(args)
  -- "Returns the solution for part two"

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Get counts for multiple paths
  knt_1_1 = self:count_trees(1, 1)
  knt_3_1 = self:count_trees(3, 1)
  knt_5_1 = self:count_trees(5, 1)
  knt_7_1 = self:count_trees(7, 1)
  knt_1_2 = self:count_trees(1, 2)
  
  -- 1. Return the solution for part two
  return knt_1_1 * knt_3_1 * knt_5_1 * knt_7_1 * knt_1_2
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------

return Trees

-- ======================================================================
-- end                      t r e e s . l u a                     end
-- ======================================================================

File: 2020/03_TobogganTrajectory/trees.js
"use strict";
// ======================================================================
// Toboggan Trajectory
//   Advent of Code 2020 Day 03 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================
exports.__esModule = true;
exports.Trees = void 0;
// ======================================================================
//                           t r e e s . t s
//
// A solver for the Advent of Code 2020 Day 03 problem
// ======================================================================
// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------
// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
var TREE = '#';
// ======================================================================
//                                                                  Trees
// ======================================================================
var Trees = /** @class */ (function () {
    function Trees(text, part2) {
        // Create a Trees object
        if (part2 === void 0) { part2 = false; }
        this.rows = 0;
        this.cols = 0;
        // 1. Set the initial values
        this.text = text === undefined ? [] : text;
        this.part2 = part2 === undefined ? false : part2;
        this.rows = 0;
        this.cols = 0;
        // 2. Process text (if any)
        if (this.text.length !== 0) {
            this.rows = this.text.length;
            this.cols = this.text[0].length;
        }
    }
    Trees.prototype.is_tree = function (col, row) {
        // Returns true if there is a tree at (col, row)
        // 1. There are no trees below the forest
        if (this.is_below(row)) {
            return false;
        }
        // 2. Check the forest square
        var mcol = col % this.cols;
        if (TREE === this.text[row][mcol]) {
            return true;
        }
        // 4. There is no spoon
        return false;
    };
    Trees.prototype.is_below = function (row) {
        // Returns true if beyond the forest
        return row >= this.rows;
    };
    Trees.prototype.count_trees = function (delta_col, delta_row) {
        // Return number of trees hit in toboggen ride
        // 1. Start at the very beginning
        var col = 0;
        var row = 0;
        var knt = 0;
        // 2. Loop until we are out of the forest
        do {
            // 3. If this is a tree, count it
            if (this.is_tree(col, row)) {
                knt += 1;
            }
            // 4. Advent to the next location
            col += delta_col;
            row += delta_row;
        } while (!this.is_below(row));
        // 5. Return the number of trees hit
        return knt;
    };
    Trees.prototype.solution = function (verbose, limit) {
        if (verbose === void 0) { verbose = false; }
        if (limit === void 0) { limit = 0; }
        if (verbose)
            console.log("solution: " + limit);
        if (this.part2) {
            // 1. Get the counts from multiple toboggan runs
            var kntOneOne = this.count_trees(1, 1);
            var kntThreeOne = this.count_trees(3, 1);
            var kntFiveOne = this.count_trees(5, 1);
            var kntSevenOne = this.count_trees(7, 1);
            var kntOneTwo = this.count_trees(1, 2);
            // 2. Return the product of the trees encountered
            return kntOneOne * kntThreeOne * kntFiveOne * kntSevenOne * kntOneTwo;
        }
        return this.count_trees(3, 1);
    };
    Trees.prototype.partOne = function (verbose, limit) {
        // Returns the solution for part one
        if (verbose === void 0) { verbose = false; }
        if (limit === void 0) { limit = 0; }
        return this.solution(verbose, limit);
    };
    Trees.prototype.partTwo = function (verbose, limit) {
        // Returns the solution for part two
        if (verbose === void 0) { verbose = false; }
        if (limit === void 0) { limit = 0; }
        // 1. Return the solution for part two
        return this.solution(verbose, limit);
    };
    return Trees;
}());
exports.Trees = Trees;
// ======================================================================
// end                      t r e e s . t s                     end
// ======================================================================
